File: src/hadoop/src/main/java/minimal_algorithms/RangeTree.java
package minimal_algorithms.hadoop;

import java.util.List;
import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;
import org.apache.avro.specific.SpecificData;
import org.apache.avro.generic.GenericRecord;
import minimal_algorithms.hadoop.statistics.StatisticsAggregator;
import minimal_algorithms.hadoop.utils.Pair;

public class RangeTree extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
    public static Schema SCHEMA$;

    public static void setSchema(Schema schema) {
        SCHEMA$ = SchemaBuilder
                .record("RangeTree").namespace("minimal_algorithms.hadoop")
                .fields()
                .name("base").type().intType().noDefault()
                .name("nodes").type().array().items(schema).noDefault()
                .endRecord();
    }

    public static Schema getClassSchema() { return SCHEMA$; }

    public Schema getSchema() { return SCHEMA$; }

    @Override
    public Object get(int field$) {
        switch (field$) {
            case 0: return BASE;
            case 1: return nodes;
            default: throw new org.apache.avro.AvroRuntimeException("Bad index");
        }
    }

    @Override
    public void put(int field$, Object value$) {
        switch (field$) {
            case 0: BASE = (Integer)value$; break;
            case 1: nodes = (StatisticsAggregator[])value$; break;
            default: throw new org.apache.avro.AvroRuntimeException("Bad index");
        }
    }

    public static RangeTree deepCopy(RangeTree tree) {
        return SpecificData.get().deepCopy(getClassSchema(), tree);
    }

    public static RangeTree deepCopy(Schema treeSchema, RangeTree tree) {
        return SpecificData.get().deepCopy(treeSchema, tree);
    }

    private int BASE;
    private StatisticsAggregator[] nodes;

    public RangeTree() {
        BASE = 0;
        nodes = new StatisticsAggregator[0];
    }

    public RangeTree(int elementsNo) {
        BASE = computeBASE(elementsNo);
        nodes = new StatisticsAggregator[2 * BASE];
    }

    public RangeTree(List<Pair<StatisticsAggregator, Integer>> elements) {
        BASE = computeBASE(elements.size());
        nodes = new StatisticsAggregator[2 * BASE];
        for (Pair<StatisticsAggregator, Integer> element : elements) {
            insert(element.getKey(), element.getValue());
        }
    }

    public void insert(StatisticsAggregator element, int start) {
        if (start < 0 || start >= BASE)
            throw new org.apache.avro.AvroRuntimeException("Position out of range: " + start);

        int pos = BASE + start;
        nodes[pos] = StatisticsAggregator.safeMerge(nodes[pos], element);
        while (pos != 1) {
            pos = pos / 2;
            nodes[pos] = StatisticsAggregator.safeMerge(nodes[2 * pos], nodes[2 * pos + 1]);
        }
    }

    public StatisticsAggregator query(int start, int end) {
        if (start > end)
            throw new org.apache.avro.AvroRuntimeException("Start (" + start + ") greater than end (" + end + ")");

        if (start < 0 || start >= BASE)
            throw new org.apache.avro.AvroRuntimeException("Position (start) out of range: " + start);

        if (end < 0 || end >= BASE)
            throw new org.apache.avro.AvroRuntimeException("Position (end) out of range: " + end);

        int vs = BASE + start;
        int ve = BASE + end;
        StatisticsAggregator result = nodes[vs];
        if (vs != ve) result = StatisticsAggregator.safeMerge(result, nodes[ve]);

        while (vs / 2 != ve / 2) {
            if (vs % 2 == 0) result = StatisticsAggregator.safeMerge(result, nodes[vs + 1]);
            if (ve % 2 == 1) result = StatisticsAggregator.safeMerge(result, nodes[ve - 1]);
            vs /= 2;
            ve /= 2;
        }
        return result;
    }

    public StatisticsAggregator[] getNodes() {
        return nodes;
    }

    public int getBASE() {
        return BASE;
    }

    private int computeBASE(int n) {
        return n <= 0 ? 0 : (int) Math.pow(2.0, Math.ceil(log2(((double) n))));
    }

    private double log2(double x) {
        return Math.log10(x) / Math.log10(2.0);
    }
}

File: src/spark/src/main/scala/minimal_algorithms/RangeTree.scala
package minimal_algorithms.spark

import minimal_algorithms.spark.statistics.StatisticsAggregator
import minimal_algorithms.spark.statistics.StatisticsUtils.safeMerge

import scala.reflect.ClassTag

/**
  * Fully binary tree. Enables searching for MAX, MIN, SUM on given range (start, end) in complexity O(n log n)
  * where n is number of initial elements.
  *
  * @param elements list of pairs (value, leaf position into which value will be inserted)
  */
class RangeTree[S <: StatisticsAggregator[S]](elements: Seq[(S, Int)])(implicit stag: ClassTag[S]) extends Serializable {
  val log2: Double => Double = (x: Double) => math.log10(x) / math.log10(2.0)
  val BASE: Int = math.pow(2.0, math.ceil(log2(elements.length.toDouble))).toInt
  var nodes: Array[S] = new Array[S](2 * BASE)
  elements.foreach{ case(element, pos) => insert(element, pos) }

  /**
    * Insert value at given leaf position.
    * @param element  object S
    * @param position [0 ... BASE - 1]
    * @return  Unit
    */
  def insert(element: S, position: Int): Unit = {
    if (position < 0 || position >= BASE)
      throw new IndexOutOfBoundsException("Position out of range: " + position)

    var pos = BASE + position
    nodes(pos) = safeMerge(nodes(pos), element)
    while(pos != 1) {
      pos = pos / 2
      nodes(pos) = safeMerge(nodes(2 * pos), nodes(2 * pos + 1))
    }
  }

  /**
    * Find aggregated value in range. start <= end
    * @param start  [0 ... BASE - 1]
    * @param end  [0 ... BASE - 1]
    * @return  Aggregated value in range.
    */
  def query(start: Int, end: Int): S = {
    if (start > end)
      throw new Exception("Start (" + start + ") greater than end (" + end + ")")
    if (start < 0 || start >= BASE)
      throw new IndexOutOfBoundsException("Position (start) out of range: " + start)
    if (end < 0 || end >= BASE)
      throw new IndexOutOfBoundsException("Position (end) out of range: " + end)

    var vs = BASE + start
    var ve = BASE + end
    var result = nodes(vs)
    if (vs != ve) result = safeMerge(result, nodes(ve))
    while (vs / 2 != ve / 2) {
      if (vs % 2 == 0) result = safeMerge(result, nodes(vs + 1))
      if (ve % 2 == 1) result = safeMerge(result, nodes(ve - 1))
      vs /= 2
      ve /= 2
    }
    result
  }
}
